В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекомендую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [1]:
from statsmodels.stats.weightstats import ztest

In [2]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [3]:
len(heights)

33

Null hypothesis H0 - there are no difference between average weight of population and random sample of students

Alternative hypothesis H1 - the average weight of random sample of students is larger 

In [4]:
null_mean = 170
alpha = 0.05

In [5]:
ztest_score, p_value = ztest(heights, value=null_mean, alternative='larger')
ztest_score, p_value

(np.float64(0.8482235371587619), np.float64(0.19815674571091746))

In [6]:
if p_value < alpha:
    print('Reject H0')
else:
    print("Fail to reject H0")

Fail to reject H0


**Conclusion** 

The data do not provide sufficient evidence, at the 5% level of significance, to conclude that the average weight of random sample of students os higher than 170.

### Завдання 2: Z-тест для двох незалежних вибірок

#### **Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [7]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [8]:
print(f'Group A mean score: {np.mean(group_a_scores):.3f} \nGroup B mean score: {np.mean(group_b_scores):.3f}')

NameError: name 'np' is not defined

Null hypothesis H0 - there are no difference between average score of group A and group B (`group_a_mean - group_b_mean = 0`)

Alternative hypothesis H1 - the average score for group A (new course) is higher than for group B (old course)

In [16]:
ztest(group_a_scores, group_b_scores, 0, alternative='larger')

(np.float64(2.357354864751502), np.float64(0.009202824276264003))

**Conclusion** 

p_value = 0.009 < 0.05 

The data provide sufficient evidence, at the 5% level of significance, to conclude that the average score for Group A, who studying new course, is higher than average score for Group B (old course). The new course is more effective. 

#### **Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [17]:
group_a_scores_small = group_a_scores[:5]
group_a_scores_small

[78.55, 72.25, 79.88, 75.0, 76.54]

In [18]:
from scipy.stats import ttest_ind

In [19]:
ttest_ind(group_a_scores_small,
          group_b_scores,
          random_state=24,
          alternative='greater')

TtestResult(statistic=np.float64(0.8167951030410814), pvalue=np.float64(0.20927397387678665), df=np.float64(43.0))

**Conclusion** 

p_value = 0.209 > 0.05 

The data do not provide sufficient evidence, at the 5% level of significance, to conclude that the average score for Group A is higher than average score for Group B.

This means that in the case of early stopping of the experiment (not enough data was collected), a false result may be obtained.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [20]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [21]:
len(before_training), len(after_training)

(30, 30)

In [30]:
mean_after = np.mean(after_training)
mean_before = np.mean(before_training)
print(f'Mean score before training: {mean_before:.3f}')
print(f'Mean score after training: {mean_after:.3f}')

Mean score before training: 50.886
Mean score after training: 54.036


In case of two related samples we could use z-test for check if difference in score after and before training higher 0.

H0: `mean_after - mean_before = 0`

H1: `mean_after - mean_before > 0`

In [ ]:
import numpy as np

In [33]:
diff_score = np.array(after_training) - np.array(before_training)
diff_score

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

In [34]:
null_mean = 0
alpha = 0.05

ztest(diff_score, value=null_mean, alternative='larger')

(np.float64(9.13890009623916), np.float64(3.1545359919175505e-20))

p_value = 3.15e-20 < 0.05 

The probability of obtaining the observed or more extreme results, assuming that the null hypothesis is true (there is no effect of training), is very small.

The data provide sufficient evidence, at the 5% level of significance, to conclude that the average score after training is higher than the average score before training.

**OR**

In the case of two related samples we can calculate a paired t-test (using `ttest_rel` from `scipy`).

This is a test for the null hypothesis that two related or repeated samples have identical average (expected) values.

In [23]:
from scipy.stats import ttest_rel

In [28]:
ttest_rel(before_training, after_training, alternative='less')

TtestResult(statistic=np.float64(-9.13890009623916), pvalue=np.float64(2.4429368127512243e-10), df=np.int64(29))

In [27]:
statistic, p_value = ttest_rel(before_training, after_training, alternative='less')
print(p_value)

2.4429368127512243e-10


**Conclusion** 

p_value = 2.4e-10 < 0.05 

The data provide sufficient evidence, at the 5% level of significance, to conclude that the average score after training is higher than average score before training.